#2373 Update on github information on mobile view, need update on the titles

irenejoeunpark opened the pull request:
## Issue This PR Addresses

#2373 More change is needed after some feedbacks on styles

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [ ] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [x] **UI**: Change which improves UI

## Description

This PR contains an update on the UI for mobile view, to include GitHub information on each post. 
When the user clicks on the GitHub icon, it expands the GitHub information of the post. 

Need feedback on the icon, design, and the idea of using Accordion.
Need to fix the title and the author name position on mobile view - need some advice on it. 


<img width="938" alt="Screen Shot 2021-11-18 at 3 20 35 PM" src="https://user-images.githubusercontent.com/54859944/142491246-caaeee93-efd3-4620-b2f0-3036c5eba49e.png">
<img width="906" alt="Screen Shot 2021-11-18 at 3 20 43 PM" src="https://user-images.githubusercontent.com/54859944/142490989-7554eba2-b59d-4214-b3ee-fd06755a79d6.png">


## Checklist

<!-- Before submitting a PR, address each item -->

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not -> Not a functional change, no need of functional tests to be added. 
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable) -> UI change, not applicable


<a href="https://gitpod.io/#https://github.com/Seneca-CDOT/telescope/pull/2491"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

DukeManh commented on src/web/src/components/Posts/Commits.tsx:
```diff
@@ -14,6 +14,10 @@ const useStyles = makeStyles((theme: Theme) =>
       lineHeight: 'normal',
       fontSize: '1.2rem',
     },
+    GitHubMobile: {
```
This class is not used anywhere right?

irenejoeunpark replied:
I removed it. Thanks for the review!

humphd commented:
I think the idea is good.  The details need some work, but having it expand vertically to reveal the GitHub info is interesting.  We need to find a way to use the space more efficiently on mobile (not jammed to the right).  Also, I'm not sure where the GitHub icon should go, or what it should look like.  It might need some text to indicate there is `More...` somehow.

I would focus on figuring out how to arrange this same info in a smaller space efficiently on mobile (e.g., what's in the revealed area).

DukeManh commented on src/web/src/components/Posts/GitHubMobile.tsx:
```diff
@@ -0,0 +1,156 @@
+      marginRight: '1rem',
+      verticalAlign: 'text-bottom',
+      paddingLeft: '6rem',
+      alignItems: 'right',
```
Did you mean `textAlign`? This property is invalid.

irenejoeunpark replied:
Fixed! Thanks :)

DukeManh commented on src/web/src/components/Posts/GitHubMobile.tsx:
```diff
@@ -0,0 +1,156 @@
+    icon: {
+      fontSize: '4rem',
+      marginRight: '1rem',
+      verticalAlign: 'text-bottom',
```
This property is invalid because `icon` is not an inline item

irenejoeunpark replied:
Fixed. Thanks!

DukeManh commented on src/web/src/components/Posts/GitHubMobile.tsx:
```diff
@@ -0,0 +1,156 @@
+      alignItems: 'right',
+    },
+    accordionSummary: {
+      width: '10 rem',
```
You have extra space between '10' and 'rem' here. It's not valid.

irenejoeunpark replied:
Fixed. Thanks!

DukeManh commented on src/web/src/components/Posts/GitHubMobile.tsx:
```diff
@@ -0,0 +1,156 @@
+      justifyContent: 'flex-end',
+    },
+    accordion: {
+      backgroundColor: 'rgba(255, 255, 255, 0)',
```
You can have it inherit main bg color: `backgroundColor: inherit`.

irenejoeunpark replied:
Thanks for the recommendation, I learned a new thing. This is fixed, thanks!

DukeManh commented on src/web/src/components/Posts/GitHubMobile.tsx:
```diff
@@ -0,0 +1,156 @@
+      backgroundColor: 'rgba(255, 255, 255, 0)',
+      border: 'none',
+      boxShadow: 'none',
+      justifyContent: 'flex-end',
```
This is not a flexbox so this won't apply.

irenejoeunpark replied:
Removed. Thanks

DukeManh commented on src/web/src/components/Posts/GitHubMobile.tsx:
```diff
@@ -0,0 +1,156 @@
+      boxShadow: 'none',
+      justifyContent: 'flex-end',
+    },
+    '&:before': {
```
This doesn't seem to belong to any class. Also, it should be `&::before` with two colons.

irenejoeunpark replied:
Removed &:before since this would affect nowhere. Thanks!

DukeManh commented on src/web/src/components/Posts/GitHubMobile.tsx:
```diff
@@ -0,0 +1,156 @@
+};
+
+const filterGitHubUrls = (urls: string[]) => {
+  const issues: Set<string> = new Set();
```
We should extract the logic for GitHub information and reuse them in both mobile and desktop components vs copying it.
Another way is to make `GithubInfo` component render the info differently using [useMediaQuery](https://mui.com/components/use-media-query/).

irenejoeunpark replied:
I am working on this, when I get all the required information from GitHubInfo, would use ```import GitHubInfo from './GitHubInfo';``` be enough for it or any other additional step is needed?

irenejoeunpark replied:
This has been updated. I removed those and reused them from GitHubInfo. Thanks

DukeManh commented on src/web/src/components/Posts/Issues.tsx:
```diff
@@ -14,6 +14,10 @@ const useStyles = makeStyles((theme: Theme) =>
       lineHeight: 'normal',
       fontSize: '1.2rem',
     },
+    GitHubMobile: {
```
You might forget to remove this class too.

irenejoeunpark replied:
Resolved!

DukeManh commented on src/web/src/components/Posts/PullRequests.tsx:
```diff
@@ -14,6 +14,10 @@ const useStyles = makeStyles((theme: Theme) =>
       lineHeight: 'normal',
       fontSize: '1.2rem',
     },
+    GitHubMobile: {
```
Same here

irenejoeunpark replied:
Fixed as well!

DukeManh reviewed (changes requested):
This is awesome.

I'm just thinking about whether we should make display the github info at the bottom of each post by default vs make users click a button and see a layout shift.

irenejoeunpark commented:
> I think the idea is good. The details need some work, but having it expand vertically to reveal the GitHub info is interesting. We need to find a way to use the space more efficiently on mobile (not jammed to the right). Also, I'm not sure where the GitHub icon should go, or what it should look like. It might need some text to indicate there is `More...` somehow.
> 
> I would focus on figuring out how to arrange this same info in a smaller space efficiently on mobile (e.g., what's in the revealed area).

I found another method for collapsable, called Colllapse in material-UI, here is an example of the usage.
https://codesandbox.io/s/77r47?file=/src/App.js:0-41
Should I use this to collapse the info horizontally? 
I am also thinking that no one would notice that the GitHub icon is for expanding information, and people would rarely click it. 
So, I am changing it to ```...``` which seems more reasonable and could lead more people to click it.

humphd reviewed (changes requested):
I think this needs more UX work.  Here's the comparison of what we have now (left) with what this does (right):

<img width="822" alt="Screen Shot 2021-11-23 at 12 38 59 PM" src="https://user-images.githubusercontent.com/427398/143076120-539e55fa-8397-4a9f-82b6-7e3f304cc14c.png">

I think adding another `...` clutters the view.  Here's my suggestion instead:

- Make the entire header (avatar logo, title, name, etc) a single clickable area
- If there is GitHub data to show, add a small GitHub icon to the right of the person's name, which would indicate that there is GitHub data to reveal
- Touching the header would open the header to a larger size

The current layout of the GitHub data isn't going to work on mobile:

<img width="394" alt="Screen Shot 2021-11-23 at 12 40 59 PM" src="https://user-images.githubusercontent.com/427398/143076369-31b59774-0466-4610-af90-2a797bafb84c.png">

We need to use the entire width somehow.  cc @Meneguini, @PedroFonsecaDEV for ideas.

irenejoeunpark commented:
> I think this needs more UX work. Here's the comparison of what we have now (left) with what this does (right):
> 
> <img alt="Screen Shot 2021-11-23 at 12 38 59 PM" width="822" src="https://user-images.githubusercontent.com/427398/143076120-539e55fa-8397-4a9f-82b6-7e3f304cc14c.png">
> 
> I think adding another `...` clutters the view. Here's my suggestion instead:
> 
> * Make the entire header (avatar logo, title, name, etc) a single clickable area
> * If there is GitHub data to show, add a small GitHub icon to the right of the person's name, which would indicate that there is GitHub data to reveal
> * Touching the header would open the header to a larger size
> 
> The current layout of the GitHub data isn't going to work on mobile:
> 
> <img alt="Screen Shot 2021-11-23 at 12 40 59 PM" width="394" src="https://user-images.githubusercontent.com/427398/143076369-31b59774-0466-4610-af90-2a797bafb84c.png">
> 
> We need to use the entire width somehow. cc @Meneguini, @PedroFonsecaDEV for ideas.

It seems like a good design. I will try to make the whole title container a clickable area and add a GitHub icon if there's any GitHub link attached.

irenejoeunpark commented:
> I think this needs more UX work. Here's the comparison of what we have now (left) with what this does (right):
> 
> <img alt="Screen Shot 2021-11-23 at 12 38 59 PM" width="822" src="https://user-images.githubusercontent.com/427398/143076120-539e55fa-8397-4a9f-82b6-7e3f304cc14c.png">
> 
> I think adding another `...` clutters the view. Here's my suggestion instead:
> 
> * Make the entire header (avatar logo, title, name, etc) a single clickable area
> * If there is GitHub data to show, add a small GitHub icon to the right of the person's name, which would indicate that there is GitHub data to reveal
> * Touching the header would open the header to a larger size
> 
> The current layout of the GitHub data isn't going to work on mobile:
> 
> <img alt="Screen Shot 2021-11-23 at 12 40 59 PM" width="394" src="https://user-images.githubusercontent.com/427398/143076369-31b59774-0466-4610-af90-2a797bafb84c.png">
> 
> We need to use the entire width somehow. cc @Meneguini, @PedroFonsecaDEV for ideas.

@humphd I have revised the code to make the title container clickable, and show the GitHub icon beside the author name when there is any GitHub info available. 

Now I faced another challenge, which the title container, which has been pinned to the top before I make the change, is now scrolled down with the post content. I tried a few different things, but it never gets pinned on top even though I only updated it to put the whole thing in the AccordionSummary. <- might be related to the issue? 
I will ask for other reviews in the slack.

AmasiaNalbandian commented:
Irene, when I click on the container to open the accordion, I believe right now that same title is a link. I can't open the containers without being navigated to the blog site. Maybe we should not use a link there?

irenejoeunpark commented on src/web/src/components/Posts/Post.tsx:
```diff
@@ -192,6 +196,31 @@ const useStyles = makeStyles((theme: Theme) =>
+      overflow: 'scroll',
+      textOverflow: 'ellipsis',
+    },
+    accordion: {
```
Thanks for the suggestion! I just tried it and it works perfectly! New knowledge gained 😉

irenejoeunpark commented:
> Irene, when I click on the container to open the accordion, I believe right now that same title is a link. I can't open the containers without being navigated to the blog site. Maybe we should not use a link there?

Hi @AmasiaNalbandian  It should be only available when your screen size is below 1025 and below in width.

irenejoeunpark commented:
> I think this needs more UX work. Here's the comparison of what we have now (left) with what this does (right):
> 
> <img alt="Screen Shot 2021-11-23 at 12 38 59 PM" width="822" src="https://user-images.githubusercontent.com/427398/143076120-539e55fa-8397-4a9f-82b6-7e3f304cc14c.png">
> 
> I think adding another `...` clutters the view. Here's my suggestion instead:
> 
> * Make the entire header (avatar logo, title, name, etc) a single clickable area
> * If there is GitHub data to show, add a small GitHub icon to the right of the person's name, which would indicate that there is GitHub data to reveal
> * Touching the header would open the header to a larger size
> 
> The current layout of the GitHub data isn't going to work on mobile:
> 
> <img alt="Screen Shot 2021-11-23 at 12 40 59 PM" width="394" src="https://user-images.githubusercontent.com/427398/143076369-31b59774-0466-4610-af90-2a797bafb84c.png">
> 
> We need to use the entire width somehow. cc @Meneguini, @PedroFonsecaDEV for ideas.

@humphd I fixed the title container to 

> > I think this needs more UX work. Here's the comparison of what we have now (left) with what this does (right):
> > <img alt="Screen Shot 2021-11-23 at 12 38 59 PM" width="822" src="https://user-images.githubusercontent.com/427398/143076120-539e55fa-8397-4a9f-82b6-7e3f304cc14c.png">
> > I think adding another `...` clutters the view. Here's my suggestion instead:
> > 
> > * Make the entire header (avatar logo, title, name, etc) a single clickable area
> > * If there is GitHub data to show, add a small GitHub icon to the right of the person's name, which would indicate that there is GitHub data to reveal
> > * Touching the header would open the header to a larger size
> > 
> > The current layout of the GitHub data isn't going to work on mobile:
> > <img alt="Screen Shot 2021-11-23 at 12 40 59 PM" width="394" src="https://user-images.githubusercontent.com/427398/143076369-31b59774-0466-4610-af90-2a797bafb84c.png">
> > We need to use the entire width somehow. cc @Meneguini, @PedroFonsecaDEV for ideas.
> 
> @humphd I have revised the code to make the title container clickable, and show the GitHub icon beside the author name when there is any GitHub info available.
> 
> Now I faced another challenge, which the title container, which has been pinned to the top before I make the change, is now scrolled down with the post content. I tried a few different things, but it never gets pinned on top even though I only updated it to put the whole thing in the AccordionSummary. <- might be related to the issue? I will ask for other reviews in the slack.

Hi, @humphd I fixed the title container to be pinned to the top of the screen. And now it looks like below. 
<img width="424" alt="Screen Shot 2021-11-23 at 11 09 57 PM" src="https://user-images.githubusercontent.com/54859944/143174803-3b703a4e-37c5-4c1a-981a-c63a612e551a.png">

And below image is when the user click to expand. 
<img width="423" alt="Screen Shot 2021-11-23 at 11 10 06 PM" src="https://user-images.githubusercontent.com/54859944/143174861-6111d643-7994-406d-89fe-2fa52f5a5e27.png">

I moved the information aligned left for a better look. 

Any feedback or suggestions are welcomed!

HyperTHD reviewed (changes requested):
Judging based on latest code, there's some stuff that I think we can fix

- There's no indication that the user has to click on the space just below the title to view the github info. You could probably try and see if you can fit like an arrow svg that can be used in the empty space allocated. Probably would want to make it work similar to a hamburger nav icon.

- You're gonna need some margin at the bottom, too much white-space at the end before the title. Also, there's some whitespace above the github info when it's opened, that's also needs to be fixed too.
![Example post](https://user-images.githubusercontent.com/16841702/143527050-28e98124-90fa-459b-9754-19bb1b07371d.png)

I think the actual github info stuff looks fine. The github user icon is a bit small, but otherwise, the rest looks fine.

humphd reviewed (changes requested):
This is getting better.  A few things:

Let's reduce the amount of whitespace at the bottom when the accordion is collapsed (compare this to what's happening on staging/production):

<img width="393" alt="Screen Shot 2021-11-26 at 9 19 21 AM" src="https://user-images.githubusercontent.com/427398/143594656-b83f23d9-81c2-429a-a494-f47a04b5286d.png">

Let's remove the line **in the mobile view** on the left, which isn't needed, since it's purpose is to divide this from the content in desktop view.  Also, let's get rid of the excessive whitespace before the GitHub info starts (again, only on mobile).

<img width="397" alt="Screen Shot 2021-11-26 at 9 17 40 AM" src="https://user-images.githubusercontent.com/427398/143594657-cd4c5afb-67d2-46de-835a-8e4b73fadabc.png">

irenejoeunpark commented:
> This is getting better. A few things:
> 
> Let's reduce the amount of whitespace at the bottom when the accordion is collapsed (compare this to what's happening on staging/production):
> 
> <img alt="Screen Shot 2021-11-26 at 9 19 21 AM" width="393" src="https://user-images.githubusercontent.com/427398/143594656-b83f23d9-81c2-429a-a494-f47a04b5286d.png">
> 
> Let's remove the line **in the mobile view** on the left, which isn't needed, since it's purpose is to divide this from the content in desktop view. Also, let's get rid of the excessive whitespace before the GitHub info starts (again, only on mobile).
> 
> <img alt="Screen Shot 2021-11-26 at 9 17 40 AM" width="397" src="https://user-images.githubusercontent.com/427398/143594657-cd4c5afb-67d2-46de-835a-8e4b73fadabc.png">

Hi @humphd, I have updated the mentioned improvements. 
Below are the images after the update
<img width="330" alt="Screen Shot 2021-11-26 at 4 39 16 PM" src="https://user-images.githubusercontent.com/54859944/143656799-3c3e290e-1b49-4b40-b268-532863f83c77.png">
<img width="326" alt="Screen Shot 2021-11-26 at 4 39 30 PM" src="https://user-images.githubusercontent.com/54859944/143656806-caa4595d-77e6-4173-8f66-b122e199df7c.png">

HyperTHD commented:
@irenejoeunpark There's still a bit of whitespace at the bottom before the first sentence of the blog post that'll need to be fixed.
![image](https://user-images.githubusercontent.com/16841702/143728988-479dd356-f2de-484b-8c3d-787d94920f1e.png)

Also, we really need a way for users to be able to know that clicking at a certain spot brings over the github info in mobile view. 
Maybe an icon? Would like to hear from everyone else about this
![image](https://user-images.githubusercontent.com/16841702/143729010-bf85a1e7-dfa6-45a0-9e0a-63fe86627a9c.png)

humphd reviewed (changes requested):
This is really getting close, nice work.

One issue I'm hitting is that it's adding overflow and scrolling in the header:


https://user-images.githubusercontent.com/427398/143775243-0fb8371d-61cf-4710-b4fa-652fa56c96e4.mov

I also want to rework the layout of the components in the accordion, but we can do that in a followup.

humphd commented:
This seems to have stalled.  @irenejoeunpark can you update us and fix the merge conflicts and review issues so that we can move forward?

irenejoeunpark commented:
> This seems to have stalled. @irenejoeunpark can you update us and fix the merge conflicts and review issues so that we can move forward?

I am working on it! I am actually stuck on detecting which one causes that scrolling thing. I know it would be with height, margin, or padding, but even if I change every single one of them it still does that.

irenejoeunpark commented:
Hi @humphd, I have tested the mobile view in [https://dev.telescope.cdot.systems/], and the issue seems like it is not only happening in this branch. 
Filed a separate issue #2583 for the fixes of the bug.

humphd commented:
Excellent, thank you.  Can you rebase to pick up the changes to `src/web/src/components/Posts/*`?

humphd commented:
This isn't perfect, but it's better.  The mechanism of showing/hiding the info is working well.  I'm not sure about the font size change on the header text.  And the layout of the inner contents are still not right for mobile.

I'll be interested to see further improvements and hear more from other reviewers.

DukeManh commented on src/web/src/components/Posts/Users.tsx:
```diff
@@ -15,6 +15,10 @@ const useStyles = makeStyles((theme: Theme) =>
       lineHeight: 'normal',
       fontSize: '1.2rem',
     },
+    GitHubMobile: {
```
This class isn't needed.

DukeManh commented on src/web/src/components/Posts/GItHubInfoMobile.tsx:
```diff
@@ -0,0 +1,51 @@
+
+  return (
+    <div>
+      <ListSubheader className={classes.root}>
```
Can you make this `<ListSubheader component="div">` and the same to every `ListSubheader` you add? We have a number of wrong uses of `ListSubheader` throughout the tree because we like its default style. By default it renders a `<li>`, `li` is meant for lists, we don't want to end up having nested `<li>`. I have fixed some `<ListSubheader component="div">` in [#2594](https://github.com/Seneca-CDOT/telescope/pull/2594) but to avoid this in the future I'll file an issue to replace all usage of `ListSubheader` with a different component while maintaining the same style.

DukeManh commented on src/web/src/components/Posts/Post.tsx:
```diff
@@ -311,62 +353,90 @@ const PostComponent = ({ postUrl, currentPost, totalPosts }: Props) => {
-                  <time dateTime={post.updated}>{`${formatPublishedDate(post.updated)}`}</time>
-                </a>
+      {!desktop && (
+        <>
```
This fragment isn't needed.

DukeManh commented on src/web/src/components/Posts/Post.tsx:
```diff
@@ -311,62 +353,90 @@ const PostComponent = ({ postUrl, currentPost, totalPosts }: Props) => {
+          </Accordion>
+        </>
+      )}
       {desktop && (
```
It'll be nice if y can combine these two expression into something like `desktop ? <A/> : <B/>`

DukeManh commented on src/web/src/components/Posts/Repos.tsx:
```diff
@@ -15,6 +15,11 @@ const useStyles = makeStyles((theme: Theme) =>
       wordWrap: 'break-word',
       lineHeight: 'normal',
     },
+    GitHubMobile: {
```
This class isn't needed.

DukeManh reviewed (changes requested):
This is getting closer. I left some reviews.

DukeManh commented on src/web/src/components/Posts/GItHubInfoMobile.tsx:
```diff
@@ -0,0 +1,51 @@
+
+const GitHubInfoMobile = ({ ghUrls }: Props) => {
+  const classes = useStyles();
+  const { repos, issues, pullRequests, commits, users } = filterGitHubUrls(ghUrls);
```
let's cache the result of this function call with useMemo so that it doesn't get called every time the `Accordion` is open. 
```suggestion
  const { repos, issues, pullRequests, commits, users } = useMemo(
    () => filterGitHubUrls(ghUrls),
    [ghUrls]
  );
```
Can you do the same for `GithubInfo`?

DukeManh commented on src/web/src/components/Posts/Post.tsx:
```diff
@@ -203,6 +209,42 @@ const useStyles = makeStyles((theme: Theme) =>
+        padding: '0 0 0 1rem',
+      },
+    },
+    gitHubIcon: {
```
You don't use a githubIcon anymore right.

DukeManh commented:
One area of the inner layout to be improved is the repos list to take up some of the horizontal space. If it looks clustered to you, feel free to design around and add visual separators between them :).
Column | Row
:-------------------------:|:-------------------------:
<img width="687" alt="Screen Shot 2021-12-16 at 11 33 09 PM" src="https://user-images.githubusercontent.com/51073515/146489552-255428c5-2753-426d-b05e-f72a0b76da61.png"> | <img width="687" alt="Screen Shot 2021-12-16 at 11 33 55 PM" src="https://user-images.githubusercontent.com/51073515/146489568-b60419c3-23cf-457c-b329-d4811bae9e26.png">

HyperTHD commented:
> One area of the inner layout to be improved is the repos list to take up some of the horizontal space. If it looks clustered to you, feel free to design around and add visual separators between them :).
> 
> Column	Row
> <img alt="Screen Shot 2021-12-16 at 11 33 09 PM" width="687" src="https://user-images.githubusercontent.com/51073515/146489552-255428c5-2753-426d-b05e-f72a0b76da61.png">	<img alt="Screen Shot 2021-12-16 at 11 33 55 PM" width="687" src="https://user-images.githubusercontent.com/51073515/146489568-b60419c3-23cf-457c-b329-d4811bae9e26.png">

I'm definitely leaning more towards the row style, given a suitable max-width ofc. Column view makes it easier to view the list of repos, at the expense of a much longer height and unused width space being used.

humphd commented:
Using row vs. column `flex` with this breakpoint could be the answer, I agree.  That looks great.

sirinoks commented:
We are closing this PR and making our own branch to revert some changes and work on the issue. I new PR will be opened once it is ready.
